### DISCUSSION TITLE
Idiomatic negative assertions [v0.4.x]

### DISCUSSION THREAD

**Discussion by jpluscplusm** [cue-lang/cue#1690] [Q&A] [answered]
**With the current language definition (0.4.x), is there an idiomatic way to perform negative assertions?**

I've fumbled my way towards this:

```
#SomeDefinition: {
	in: !~"[A-Z]"
}

positive_test: #SomeDefinition & {
	in: "abc"
}

negative_test: true
negative_test: (
		#SomeDefinition & {
		in: "ABC"
	}) == _|_
```

... but it'd be hard to describe it as "elegant" :-)

Also, AFAICT I can't wrap this in an abstracting construct such as this:

```
#SomeDefinition: {
	in: !~"[A-Z]"
}

negative_test: #TestNotAllowed & {
	test: #SomeDefinition & {
		in: "ABC"
	}
}

#TestNotAllowed: {
	test:   _
	result: test == _|_
	result: true
}
```

... because the act of assembling `#TestNotAllowed`'s input param, `test`, causes `_|_` to be reached, and evaluation to stop.

The best I've managed is this:

```
#SomeDefinition: {
	in: !~"[A-Z]"
}

negative_test: #TestNotAllowed & {
	a: #SomeDefinition
	b: {
		in: "ABC"
	}
}

#TestNotAllowed: {
	a:      _
	b:      _
	result: (a & b) == _|_
	result: true
}
```

This works, but leaves an incomplete value in `negative_test.a`, which means that I can only run this test with `cue vet` and NOT `cue vet -c`. This means I can't combine any such abstracted negative tests with tests that assert completeness - requiring at minimum a double pass over any test suite, and the associated downstream combinatorial complexity that that implies.

I know language changes such as `must()` and `not()` are coming Real Soon Now [;-)], so I'm more than happy to make do with the first, inelegant version of these tests that I noted at the top of this discussion.

But: am I missing a trick? What are other folks doing? ❤️

**Comment by myitcv** [comment, marked as answer]
I think you're reaching for https://github.com/cue-lang/cue/issues/575 and https://github.com/cue-lang/cue/issues/571 (both linked from https://github.com/cue-lang/cue/issues/943)
